[RFC] Testing Infrastructure

claussni opened the pull request:
**Integration test with /sword resource fails because of Spring bean initialization error**
- How does testing work with F4?
- What are the dependencies?
- Why we need to manually require a specific Jersey version to make F4 run?

claussni committed:
Return plain Abdera Service Document

Adds integration test infrastructure

awoods commented:
A couple of pre-code-review comments:
- Running `mvn clean install` fails if an application is already using port 8080
  *\* See: https://github.com/fcrepo4/fcrepo4/blob/master/pom.xml#L704
- Running `mvn clean install` still fails even if nothing is on port 8080 on the following integration test:

``` bash
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.146 sec <<< FAILURE! - in org.fcrepo.sword.integration.SWORDProviderIT
returnsAtomServiceDocumentMediaType(org.fcrepo.sword.integration.SWORDProviderIT)  Time elapsed: 0.042 sec  <<< FAILURE!
java.lang.AssertionError: Expected different status code expected:<200> but was:<500>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:743)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.junit.Assert.assertEquals(Assert.java:555)
        at org.fcrepo.sword.integration.SWORDProviderIT.assertStatusCode(SWORDProviderIT.java:75)
        at org.fcrepo.sword.integration.SWORDProviderIT.returnsAtomServiceDocumentMediaType(SWORDProviderIT.java:70)
```

claussni committed:
Use build-helper-maven-plugin

Reserves a port for integration testing.

claussni commented:
I included the build-helper-maven-plugin. Oddly, it doesn't seem to work with `Fedora 4.1.0 BOM`. Needed to upgrade to `LASTEST`.

awoods commented on src/main/java/org/fcrepo/sword/http/SWORDServiceResource.java:
```diff
@@ -28,15 +31,25 @@
  */
 @Scope("request")
 @Path("/sword")
```
Using the `@Path` of `/sword` will conflict with the FedoraLdp REST path, and would otherwise pose problems for users who wanted to create a top-level `/sword` repository resource.
Sticking with the convention of `/fcr:sword` would be safe. Would using `/fcr:sword` be reasonable from your perspective?

claussni replied:
Absolutely. I will change that.

I built this similar to the OAI provider which utilizes `/oai` without the `fcr` namespace. Is the OAI provider different from a SWORD provider?

awoods replied:
No, the OAI provider needs to be fixed as well... the OAI provider issue is what made me also notice it here.

awoods commented on src/test/java/org/fcrepo/sword/http/SWORDServiceResourceTest.java:
```diff
@@ -13,24 +13,38 @@
-public class AbstractSWORDProviderIT {
+@ContextConfiguration("/spring/sword.xml")
+@WebAppConfiguration
+public class SWORDServiceResourceTest {
```
The servlet container is being spun up for this test, so I would still consider it an "integration test", and hence keep the class name as `SWORDServiceResourceIT`.
Unit tests (ClassTest.java) usually take the form of creating a `new` object and exercising its methods.

claussni replied:
That's not intended. This should clearly be a unit test. I tried to use the Spring DI container but the problem here is that there is no scope `request`. Setting the scope of `SWORDServiceResource` to `prototype` resolves this.

awoods commented on src/main/java/org/fcrepo/sword/http/SWORDServiceResource.java:
```diff
@@ -28,15 +31,25 @@
  */
 @Scope("request")
```
Do you need the `sword` service to create a new instance of this class with every request? or can the class be a singleton within the container? Setting the `@Scope` to `prototype` allows your current tests to pass successfully.
If you need `Scope("request")` we will need to dig into why that is not currently working.

claussni replied:
Again, I went with the OAI Provider example here.  But `request` isn't actually needed. I checked the Spring MVC documentation on scopes again and there is no reason for this service not to be a singleton. This [StackOverflow question](http://stackoverflow.com/questions/16795303/must-spring-mvc-classes-be-thread-safe) makes it pretty clear.

Why `@Scope("request")` isn't working still remains a question though.

claussni committed:
Build against Fedora 4.1.1 instead of latest snapshot

claussni committed:
Use singleton scope instead of request scope

Request scope causes problems with unit testing because the scope
`request` is not valid outside of a web context. Further, Fedora
integration testing doesn't work properly with this scope yet.

Singleton scope is fine because the SWORDServiceResource is ment to be
stateless. Multi threading is the default state of operation for the
Spring dispatcher anyway, so there is no need for request scoping.

claussni committed:
Use fcr namespace for resource path

https://github.com/fcrepo4-labs/fcrepo4-swordserver/pull/1#discussion_r28203556

claussni commented:
Thanks for the comments @awoods! I hope I can go further with this testing issues out of the way.
